src/tools/ranges/TransformView.hpp

```diff
@@ -0,0 +1,119 @@
+#ifndef NJOY_TOOLS_RANGES_TRANSFORMVIEW
+#define NJOY_TOOLS_RANGES_TRANSFORMVIEW
+
+// system includes
+#include <functional>
+#include <iterator>
+
+// other includes
+#include "tools/ranges/ViewBase.hpp"
+#include "tools/ranges/IteratorView.hpp"
+#include "tools/ranges/make_view.hpp"
+
+namespace njoy {
+namespace tools {
+namespace ranges {
+
+/**
+ *
+ *  @brief A view that applies a transformation on every element of a range
+ */
+template < typename Range, typename Transform >
+class TransformView : public ViewBase< TransformView< Range, Transform > > {
+
+  /* type aliases */
+  using BaseIterator = decltype( std::cbegin( std::declval< Range& >() ) );
+
+  /* fields */
+  Range base_;
+  Transform transform_;
+
+public:
+
+  #include "tools/ranges/TransformView/Iterator.hpp"
+
+  using value_type = typename std::iterator_traits< Iterator >::value_type;
+  using difference_type = typename std::iterator_traits< Iterator >::difference_type;
+  using size_type = std::size_t;
+  using pointer = typename std::iterator_traits< Iterator >::pointer;
+  using reference = typename std::iterator_traits< Iterator >::reference;
+  using const_reference = const reference;
+  using iterator = Iterator;
+  using const_iterator = const Iterator;
+
+  /* constructor */
+
+  /**
+   *  @brief Default constructor
+   */
+  constexpr TransformView() = default;
+
+  /**
+   *  @brief Constructor
+   *
+   *  @param[in] range        the range to be transformed
+   *  @param[in] transform    the transformation to be applied to the range
+   */
+  constexpr TransformView( Range range, Transform transform ) :
```

Comment:
Looked it up in cppref - I didn't know transform_view copied the data, I would have assumed it made a view of the data first.   Is there any reason to not store an IteratorView of the range instead of copying the range itself?  When I see a 'make_foo" function I expect it to basically forward along the args to the ctor of Foo.  But in this case TransformView's ctor has different behavior than make_transform_view.

Reply:
I was as surprised as you were when I saw that. ranges v3 actually does create a subrange (i.e. IteratorView) for the transport_view. I was actually debating on putting an IteratorView into the TransportView to avoid copies.